:stopinsert should leave terminal-mode

glacambre opened the pull request:
This PR closes issue #7807.

I tried to implement a test in d40968e but couldn't get it to fail when this patch is not applied (even though the bug can still be reproduced when using the steps described in #7807). I'd love to have help figuring what I'm doing wrong here.

justinmk commented on test/functional/terminal/api_spec.lua:
```diff
@@ -66,5 +66,47 @@ describe('api', function()
+
+  it("stopinsert RPC request during insert-mode", function()
+    -- Start the socket from the child nvim.
+    child_session.feed_data(":echo serverstart('"..socket_name.."')\n")
```
`child_session.feed_data` is a hack for when we need to test a child Nvim (i.e. `nvim` running inside `:terminal`). I don't think that's needed for testing this change, using a regular session will avoid a lot of complexity.

I would move the test to `test/functional/terminal/edit_spec.lua` and probably the screen tests aren't needed, only need to check `nvim_get_mode`.

glacambre replied:
I see, I should have spent more time trying to figure out what I was doing instead of blindly copy/pasting what I saw.

Thanks for the explanation :)

glacambre committed:
:stopinsert should leave terminal-mode #9856

Problem:  Calling :stopinsert from RPC while in terminal-mode does not
          go back to normal-mode.
Solution: Implement a check() handler for state_enter(), adapted from
          insert_check().

Fix #7807

justinmk commented on test/functional/terminal/ex_terminal_spec.lua:
```diff
@@ -131,6 +131,14 @@ describe(':terminal (with fake shell)', function()
+    command(':terminal')
+    feed("i[tui] insert-mode")
+    eq({ blocking=false, mode='t' }, nvim('get_mode'))
+    command('stopinsert')
```
This fails correctly before your patch, because all of the `test/functional` tests run from a Lua RPC client to Nvim. So `command()` makes an RPC call.

justinmk commented:
@glacambre I pushed an update to this PR with an updated test.

Thanks for debugging this, it's nice to have this fixed.
